### PR TITLE
Add packet-coalescing algorithm for UDP endpoints

### PR DESCRIPTION
**Pull request by jkflying**
This is a packet coalescing implementation for UDP endpoints using a default of 1ms/1200B transmit thresholds. Before, especially when communicating over ethernet, each (on average) 40B Mavlink packet could end up in its own IP packet, amplifying traffic significantly.

I've tested that it still works for UDP, and I haven't noticed any performance regressions, but also haven't verified that it correctly coalesces UDP packets. Unit tests would be nice.

General design is:

- On receipt of data to send, push it into the TX buffer
- If there is no send event scheduled, schedule one to happen in 1ms
- If the buffer has more data than 1200B, write immediately

Theoretically, setting the send size threshold to 1 should give exactly the same performance as before, minus a single extra epoll wakeup on start.

### CODE REVIEW

**Comment by jkflying** [comment]
@caleridas I noticed a valgrind issue I introduce here on SIGINT, I guess this will also affect the unit tests in the teardown:
```
^C
==1924== Invalid write of size 1
==1924==    at 0x11C1B5: Mainloop::del_timeout(Timeout*) (mainloop.cpp:687)
==1924==    by 0x11539E: UdpEndpoint::~UdpEndpoint() (endpoint.cpp:735)
==1924==    by 0x1153E8: UdpEndpoint::~UdpEndpoint() (endpoint.cpp:738)
==1924==    by 0x11CEF9: Mainloop::free_endpoints(options*) (mainloop.cpp:576)
==1924==    by 0x10C5DE: main (main.cpp:879)
==1924==  Address 0x5fff03c is 12 bytes inside a block of size 64 free'd
==1924==    at 0x4C3123B: operator delete(void*) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==1924==    by 0x11F14E: Mainloop::loop() (mainloop.cpp:366)
==1924==    by 0x10C5CF: main (main.cpp:877)
==1924==  Block was alloc'd at
==1924==    at 0x4C3017F: operator new(unsigned long) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==1924==    by 0x11C075: Mainloop::add_timeout(unsigned int, std::function<bool (void*)>, void const*) (mainloop.cpp:646)
==1924==    by 0x115A5D: UdpEndpoint::UdpEndpoint() (endpoint.cpp:716)
==1924==    by 0x11C744: Mainloop::add_endpoints(Mainloop&, options*) (mainloop.cpp:506)
==1924==    by 0x10C348: main (main.cpp:874)
```

There's an attempted double-free, and this is because the main loop tries to clean up all of the timers instead of the endpoint that requested the timer being responsible for cleaning their own timers up. The intended design pattern isn't exactly clear to me, since in the Endpoint it cleans up its own timer, do you know what was intended?

Edit: I think the solution in https://github.com/Auterion/mavlink-router/pull/20/commits/837d83ccbb154d759a1d6d0d00cf035822e048a8 will work, but it doesn't really solve the architectural problem of who owns the timeouts.

File: src/mavlink-router/endpoint.cpp [hunk 1]
```diff
@@ -326,8 +326,8 @@ bool Endpoint::accept_msg(int target_sysid, int target_compid, uint8_t src_sysid
     if (has_sys_comp_id(src_sysid, src_compid))
         return false;
 
-    if (msg_id != UINT32_MAX && 
-        _message_filter.size() > 0 && 
+    if (msg_id != UINT32_MAX &&
```
Comment (caleridas) [inline src/mavlink-router/endpoint.cpp:329]: please separate commit for unrelated cleanups

File: src/mavlink-router/endpoint.cpp [hunk 2]
```diff
@@ -771,22 +796,45 @@ ssize_t UdpEndpoint::_read_msg(uint8_t *buf, size_t len)
 
 int UdpEndpoint::write_msg(const struct buffer *pbuf)
 {
+    if (tx_buf.len + pbuf->len > TX_BUF_MAX_SIZE)
```
Comment (caleridas) [inline src/mavlink-router/endpoint.cpp]: line-wrapping wrt { looks inconsistent, see also below

File: src/mavlink-router/endpoint.cpp [hunk 3]
```diff
@@ -771,22 +796,45 @@ ssize_t UdpEndpoint::_read_msg(uint8_t *buf, size_t len)
...
+    {
+        log_debug("Dropping message, tx buffer full");
+        return 0;
+    }
+    memcpy(&tx_buf.data[tx_buf.len], pbuf->data, pbuf->len);
+    tx_buf.len += pbuf->len;
+
```
Comment (caleridas) [inline src/mavlink-router/endpoint.cpp:794]: this means you would hypothetically exceed the specified MTU size by up to the amount of a mavlink frame (minus one) -- you need to check for overflow and force flush before appending

structure that I would suggest: two functions:
- force_write(): performs actual write, deschedules timer (if there was any pending) to avoid early send of following data
- schedule_delayed_write(deadline): schedule write to happen latest at deadline of all buffer contents; leaves present timeout if it has earlier deadline, otherwise updates; post-condition: timer scheduled

File: src/mavlink-router/endpoint.cpp [hunk 4]
```diff
@@ -771,22 +796,45 @@ ssize_t UdpEndpoint::_read_msg(uint8_t *buf, size_t len)
...
+    if (tx_buf.len >= _nagle_buf_max_packet_size)
+    {
+        int bytes_remaining = _write_internal_buffer();
+        if (bytes_remaining > 0)
+        {
+            _reschedule_write();
+        }
+    } else if (!_write_scheduled) {
+        _reschedule_write();
```
Comment (caleridas) [inline src/mavlink-router/endpoint.cpp]: the check for "_write_scheduled" and only conditionally calling "_reschedule_write" if false is not consistent

would suggest moving check into method -- so after this is called, write is scheduled

the most logic thing would be to have the deadline as parameter to the schedule call -- that way you can have different deadline based on packet class (see lorenz' comments)
the schedule function would then compare with "presently" set deadline from previous packets and reschedule to pick earliest ones
may want to support parameterizing this on maximum delay

File: src/mavlink-router/endpoint.cpp [hunk 5]
```diff
@@ -795,17 +843,21 @@ int UdpEndpoint::write_msg(const struct buffer *pbuf)
...
-        log_debug("Discarding packet, incomplete write %zd but len=%u", r, pbuf->len);
-    }
+    tx_buf.len = std::max(ssize_t(0), (ssize_t)tx_buf.len - r);
+    // memcpy isn't safe for overlapping regions
+    memmove(tx_buf.data, &tx_buf.data[r], tx_buf.len);
 
     log_debug("UDP: [%d] wrote %zd bytes", fd, r);
 
-    return r;
+    return tx_buf.len;
```
Comment (caleridas) [inline src/mavlink-router/endpoint.cpp]: I see nowhere where the pending timeout is cancelled in case the write happens due to buffer being to full -- see other comment regarding code structure recommendation

File: src/mavlink-router/endpoint.h [hunk 6]
```diff
@@ -185,6 +185,14 @@ class UdpEndpoint : public Endpoint {
     struct sockaddr_in sockaddr;
 
 protected:
+
+    void _reschedule_write();
+    int _write_internal_buffer();
+    bool _write_scheduled;
+
+    Timeout* _write_schedule_timer = nullptr;
+    const unsigned int _nagle_buf_max_packet_size, _nagle_max_timeout_ms;
```
Comment (caleridas) [inline src/mavlink-router/endpoint.h]: nagle is a misnomer, that algorithm has no timers but relies on ACKS which is not applicable -- suggest just dropping the term it is slightly misleading

File: src/mavlink-router/endpoint.h [hunk 7]
```diff
@@ -199,7 +199,7 @@ class UdpEndpoint : public Endpoint {
 class TcpEndpoint : public Endpoint {
 public:
     TcpEndpoint();
-    ~TcpEndpoint();
+    virtual ~TcpEndpoint();
```
Comment (caleridas) [inline src/mavlink-router/endpoint.h]: please use override instead of virtual

File: src/mavlink-router/mainloop.cpp [hunk 8]
```diff
@@ -683,7 +683,7 @@ void Mainloop::set_timeout(Timeout *t, uint32_t timeout_msec)
 
 void Mainloop::del_timeout(Timeout *t)
 {
-    if (t) {
+    if (_timeouts && t) {
```
Comment (caleridas) [inline src/mavlink-router/mainloop.cpp]: maybe that fix should go first in sequence? I think the other code will crash without this

**Review by caleridas — CHANGES_REQUESTED** [review]
Looks okay, a couple of corner cases

Regarding testing, I just realized that https://github.com/Auterion/mavlink-router/pull/16 is upstream but not in auterion -- provides testability for stuff involving mainloop, could you look into using that to verify that it behaves and also the corner cases that I pointed out in review are met? (over-sized frames, forgotten timer cancel leading to "early" send of subsequent stuff etc.)

File: src/mavlink-router/endpoint.cpp [hunk 9]
```diff
@@ -795,17 +843,21 @@ int UdpEndpoint::write_msg(const struct buffer *pbuf)
...
+    _stat.write.bytes += r;
 
-    /* Incomplete packet, we warn and discard the rest */
-    if (r != (ssize_t) pbuf->len) {
-        _incomplete_msgs++;
-        log_debug("Discarding packet, incomplete write %zd but len=%u", r, pbuf->len);
-    }
+    tx_buf.len = std::max(ssize_t(0), (ssize_t)tx_buf.len - r);
+    // memcpy isn't safe for overlapping regions
+    memmove(tx_buf.data, &tx_buf.data[r], tx_buf.len);
```
Comment (caleridas) [inline src/mavlink-router/endpoint.cpp:834]: partial sends of frames is not something that can practically happen -- it means either you are violating the MTU of the protocol (logic error, you should terminate program), or the kernel has run out of memory (then you would have worse problems, and you should also terminate program)

possibly partial sends would also violate the deadline of the data that has been buffered, so there is little sense

File: src/mavlink-router/endpoint.cpp [hunk 10]
```diff
@@ -707,9 +707,34 @@ int UartEndpoint::add_speeds(std::vector<unsigned long> bauds)
 }
 
 UdpEndpoint::UdpEndpoint()
```
Comment (caleridas) [inline src/mavlink-router/endpoint.cpp:699]: packet coalescing with delays should probably only apply to really "outgoing" network

we are using udp also for internal traffic (yes we should not, that should all be UDS), but for the time being internal traffic should not be subject to delays

**Review by caleridas — CHANGES_REQUESTED** [review]
additional comments that I forgot, sorry

**Comment by caleridas** [comment]
Regarding pull description: the 35x amplification is a gross exaggeration -- the packet is not filled up to 1500 bytes mtu -- ip+udp headers makes up 32 bytes so you get ~ 2x amplification from this (and then admittedly vpn nonsense comes in and likely adds another 64, so you end up with 4x amplification in absolutely worst case).

At 1ms coalescing delay and high-speed uart full throughput of 1.5MBit you will get at most ~190 bytes within one 1ms, so effectiveness of this as implemented is also pretty limited.

While I agree that many small packets are not nice and should be avoided, neither full impact nor effectiveness is so clear and I would be cautious overstating it.

File: src/mavlink-router/endpoint.cpp (hunk 3 above)
Comment (jkflying) [reply src/mavlink-router/endpoint.cpp:794]: I guess the question is whether it is better to have a 1-message-size conservative threshold for triggering force_write(), and send the packets immediately, or to get the little bit of extra throughput but at the cost of always having that one extra packet being delayed by up to the coalescing time.

I was a little unhappy with the function naming, this is a nicer abstraction of the problem.

File: src/mavlink-router/endpoint.cpp (hunk 4 above)
Comment (jkflying) [reply src/mavlink-router/endpoint.cpp]: I'd prefer to work with deadlines too, but that would require more drastic mavlink-router surgery. I think we can work reasonably with the delay-based API just by doing a force_write() in the case where a high-priority message comes through. Yes, we lose the coalescing capabilities for that message type, but I generally see two classes of messages: general telemetry streaming which isn't sensitive to an extra 20ms of latency, and command-ack pairs which are sensitive to latency. I can't think of a use-case for in between.

File: src/mavlink-router/endpoint.cpp (hunk 5 above)
Comment (jkflying) [reply src/mavlink-router/endpoint.cpp]: There isn't API support in mavlink-router for cancelling timers at the epoll level. I'm not sure if it makes sense to add that, or if I should just ignore the wakeup if I didn't want the wakeup. Thoughts?

File: src/mavlink-router/endpoint.h (hunk 6 above)
Comment (jkflying) [reply src/mavlink-router/endpoint.h]: Will do

File: src/mavlink-router/endpoint.cpp (hunk 10 above)
Comment (jkflying) [reply src/mavlink-router/endpoint.cpp:699]: Agreed. Before merging this I'd like to add the capability of configuring the value per-endpoint. For radio links, having a delay of less than ~20ms probably won't help that much for typical traffic. For internal of course we don't want any delay at all.

**Comment by jkflying** [comment]
>Regarding pull description: the 35x amplification is a gross exaggeration -- the packet is not filled up to 1500 bytes mtu -- ip+udp headers makes up 32 bytes so you get ~ 2x amplification from this (and then admittedly vpn nonsense comes in and likely adds another 64, so you end up with 4x amplification in absolutely worst case).

>At 1ms coalescing delay and high-speed uart full throughput of 1.5MBit you will get at most ~190 bytes within one 1ms, so effectiveness of this as implemented is also pretty limited.

Many radios (eg. microhard) are half duplex, so MTU sharing doesn't work. Admittedly video is the major downlink data-hog, and uplink is pretty sparse, but we've seen issues with uplink being starved due to downlink saturation in degraded signal conditions. At 1ms maximum delay I agree there isn't much gain, but for groundstations over a radio link 20ms of coalescing could free occasional frames to let uplink data through.

File: src/mavlink-router/endpoint.cpp (hunk 9 above)
Comment (jkflying) [reply src/mavlink-router/endpoint.cpp:834]: I was worrying about if someone dumps a giant amount of UDP over the internal interface faster than the link layer can sink it, but I guess in that case dropping data is the right thing to do anyway.

File: src/mavlink-router/endpoint.cpp (hunk 5 above)
Comment (caleridas) [reply src/mavlink-router/endpoint.cpp]: if it is not possible to cancel then you would at least need to add code to ignore, so you would need to store "correct" deadline somewhere in order to distinguish planned/unplanned wakeups

not doing it means you will get early wakeups from dangling timers and will send prematurely, thus partly losing the effect of coalescing (maybe that is part of what is also observed in wireshark tests)

File: src/mavlink-router/endpoint.cpp (hunk 4 above)
Comment (caleridas) [reply src/mavlink-router/endpoint.cpp]: see other comment, if you cannot cancel relative timers, then you need to track deadline anyways to detect premature wakeups

that means you would need to track "absolute send deadline" in the endpoint class, while communicating "relative timeout" to the mainloop class without requiring major surgery there

I don't think this changes fundamentally when deadline is always either "immediate" or "now + 20ms"

**Comment by nicovanduijn** [comment]
As stated in our internal discussion, I do agree with coalescing in principle (How can anyone not?), but I do want to raise to points:

1) We can't and shouldn't do coalescing for commands and ack's, at least until the gimbal interface is implemented in a more meaningful matter. Otherwise this will make the gimbal unusably slow. (We are currently sending commands and waiting for their acks at high rate, sending continuous position commands... This is a separate discussion, but I think it's taking shape)

2) I'm not convinced this does anything significant to the bandwidth consumed by mavlink in its current state. With video streaming disabled, I get the following mavlink traffic:
image

Compare to the "old" mavlink router:
image

So there is some coalescing happening, but not much

File: src/mavlink-router/endpoint.cpp (hunk 3 above)
Comment (caleridas) [reply src/mavlink-router/endpoint.cpp:794]: not sure I understand the first paragraph correctly, but to clarify what I was trying to say: if something is called "buf_max_packet_size", then I would be slightly surprised by behavior that the *real* max packet size is in reality "buf_max_packet_size + X" where X is size of mavlink frame -- it might lead to accidental udp packet fragmentation if someone goes in knowing the "correct" MTU and assumes the code behaves. My comment is about "adhering to what is promised in the name". I don't think you "gain" efficiency but being fuzzy about the API, you just artificially increase the max packet size by an unspecified amount.

**Comment by LorenzMeier** [comment]
@jkflying I want to honor your time off, so we'll try to take it forward here in the meantime. Thanks!

**Comment by dogmaphobic** [comment]
> 1. at least until the gimbal interface is implemented in a more meaningful matter.

This has been in the works for quite some time:
https://docs.google.com/document/d/10aNb1Gp4FWcin4g22L5hEdGbCe8EGi3YMmtltHm4X20/edit

For Vantage for example, I disabled waiting for ACKs altogether and I am treating those commands as messages. It will likely stay that way until we move to the new protocol.

> 2\. I'm not convinced this does anything significant to the bandwidth consumed by mavlink in its current state.

I assume this is using the 1ms trigger? I run into the same from QGC's side. There, the traffic is so sparse to the point of not making a great deal of sense to coalesce. I would need to wait at least 67ms to be able to get two messages into one send. QGC is somewhat constant though. On the other hand, the FC data rate varies quite a bit depending on the link and configuration.

File: src/mavlink-router/endpoint.cpp (hunk 3 above)
Comment (jkflying) [reply src/mavlink-router/endpoint.cpp:794]: Ah, yes, my bad. I earlier had an extra variable as `trigger' which was one packet size smaller, but it was removed in a refactoring pass. 

I'll make the behavior reflect the naming.

File: src/mavlink-router/endpoint.cpp (hunk 5 above)
Comment (jkflying) [reply src/mavlink-router/endpoint.cpp]: From my understanding of [`timerfd_settime`](https://linux.die.net/man/2/timerfd_settime), it only tracks a single timeout for any given `fd`, and drops the old timeout if that `fd` had one registered. So we shouldn't have dangling wakeups. However, digging into mavlink-router it seems it configures the timer to wakeup every x milliseconds forever, not just one-shot. I'll disable the timer when it isn't needed.

**Comment by jkflying** [comment]
Closed in favor of #23
